### PR TITLE
Gaze controller upgraded

### PR DESCRIPTION
**Pull request by plinioMoreno**
Gaze controller was upgraded, matching the current version of the iCub gaze controller

### CODE REVIEW

File: vizzy_yarp_icub/src/libraries/fKin/vizzyFwd.cpp [hunk 1]
```diff
@@ -200,6 +200,71 @@ void vizzyEyeNeckRef::allocate(const string &_type)
...
+    Matrix H0(4,4);
+    H0.zero();
+    if (_root_link=="base_link")
+    {
+    //1.0 0.0 0.0 0.189861 0.0 0.000004 -1.0 0.0 -0.0 1.0 -0.000004 0.535797 0.0 0.0 0.0 1.0)
+        H0(0,0)=1.0;
+        H0(1,2)=-1.0;
+        H0(2,1)=1.0;
+        H0(3,3)=1.0;
+    H0(0,3)=0.189861;
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/libraries/fKin/vizzyFwd.cpp]: indentation

File: vizzy_yarp_icub/src/libraries/fKin/vizzyFwd.cpp [hunk 2]
```diff
@@ -200,6 +200,71 @@ void vizzyEyeNeckRef::allocate(const string &_type)
...
+    else if (_root_link=="waist")
+    {
+        H0(0,0)=1.0;
+        H0(1,2)=-1.0;
+        H0(2,1)=1.0;
+        H0(3,3)=1.0;
+    }
+    setH0(H0);
+
+#if 0
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/libraries/fKin/vizzyFwd.cpp]: this `#if 0` block can be removed

File: vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/controller.cpp [hunk 3]
```diff
@@ -1,8 +1,8 @@
 /* 
  * Copyright (C) 2010 RobotCub Consortium, European Commission FP6 Project IST-004370
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/controller.cpp]: copyright 2018 vislab

File: vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/controller.cpp [hunk 4]
```diff
@@ -864,6 +864,7 @@ void Controller::run()
         if (commData->neckPosCtrlOn)
         {
             posdeg=(CTRL_RAD2DEG)*IntPlan->get();
+            //yInfo("Curr neck 0 %f, 1 %f",posdeg.data()[0],posdeg.data()[1]);
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/controller.cpp]: remove commented debug line

File: vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp [hunk 5]
```diff
@@ -191,7 +191,7 @@ class HeadCenter_NLP : public Ipopt::TNLP
     bool eval_f(Ipopt::Index n, const Ipopt::Number* x, bool new_x,
                 Ipopt::Number& obj_value)
     {
-        obj_value=0.0;
+        /*obj_value=0.0;
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp]: please remove

File: vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp [hunk 6]
```diff
@@ -200,16 +200,24 @@ class HeadCenter_NLP : public Ipopt::TNLP
...
     bool eval_grad_f(Ipopt::Index n, const Ipopt::Number* x, bool new_x,
                      Ipopt::Number* grad_f)
     {
+        /*for (Ipopt::Index i=0; i<n; i++)
+            grad_f[i]=x[i]-qRest[i];*/
+        computeQuantities(x);
         for (Ipopt::Index i=0; i<n; i++)
-            grad_f[i]=x[i]-qRest[i];
+        {
+            grad_f[i]=(dot(AnaJacobZ,i,Hxd,3)+dot(Hxd,2,GeoJacobP,i))/mod
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp]: please document Vizzy-specific optimization

File: vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp [hunk 7]
```diff
@@ -218,11 +226,11 @@ class HeadCenter_NLP : public Ipopt::TNLP
     bool eval_g(Ipopt::Index n, const Ipopt::Number* x, bool new_x, Ipopt::Index m,
                 Ipopt::Number* g)
     {
-        computeQuantities(x);
+        /*computeQuantities(x);
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp]: please remove commented lines, add Vizzy-specific comments

File: vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp [hunk 8]
```diff
@@ -232,7 +240,7 @@ class HeadCenter_NLP : public Ipopt::TNLP
                     Ipopt::Index m, Ipopt::Index nele_jac, Ipopt::Index* iRow,
                     Ipopt::Index *jCol, Ipopt::Number* values)
     {
-        if (!values)
+        /*if (!values)
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp]: please remove commented lines

File: vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp [hunk 9]
```diff
@@ -342,7 +349,7 @@ Vector GazeIpOptMin::solve(const Vector &q0, Vector &xd, const Vector &gDir)
     yInfo("3");
     nlp->set_bound_inf(lowerBoundInf,upperBoundInf);
     yInfo("4");
-    nlp->setGravityDirection(gDir);
+    //nlp->setGravityDirection(gDir);
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/gazeNlp.cpp]: please explain why no gravity compensation on Vizzy

File: vizzy_yarp_icub/src/libraries/fKin/vizzyFwd.cpp [hunk 10]
```diff
@@ -77,8 +77,8 @@ void vizzyEye::allocate(const string &_type, const std::string &_root_link)
         setH0(H0);
 
       //pushLink(new iKinLink(  R,      D,      alpha,     theta,       minAngle,         maxAngle));
-
-    if (iKinLimb::getType()=="left")
+    yInfo(iKinLimb::getType());
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/libraries/fKin/vizzyFwd.cpp]: please verify if yInfo print was removed

File: vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/solver.cpp [hunk 11]
```diff
@@ -770,17 +786,17 @@ void Solver::run()
         gDir[2]=gDir[3]=1.0;
         if (commData->stabilizationOn)
         {
-            Vector acc=commData->get_imu().subVector(3,5); 
+            Vector acc=commData->get_imu().subVector(3,5);
             acc.push_back(1.0); // impose homogeneous coordinates
 
-            Matrix H=imu->getH(cat(fbTorso,fbHead.subVector(0,2)));
-            gDir=H*acc;            
+            Matrix H=imu->getH(cat(fbTorso,fbHead.subVector(0,1)));
```
Comment (gsaponaro) [inline vizzy_yarp_icub/src/modules/vizzy_iKinCtrlGaze/solver.cpp]: Vizzy neck: 2 degrees

**Comment by plinioMoreno** [comment]
Thanks @gsaponaro!

**Pull request merged**